### ISSUE TITLE
Support multiple datasets of different lengths

### ISSUE THREAD

**Issue by andrrizzi** [luigibonati/mlcvs#47]
Currently, ``FastDictionaryLoader`` support multiple datasets only if they have the same number of samples. We should look into whether it's possible to remove this limitation as it is probably the most common case.

**Comment by andrrizzi** [comment]
A couple of things we need to discuss about this feature:
1. Should we support multiple batch sizes?
2. How do we define the number of batches in an epoch (i.e. ``len(dataloader)``) when the datasets have different size?
   Option 1: The size of the smallest dataset. This means we're not guaranteed to see all the data of the large dataset.
   Option 2: The size of the largest dataset. For this, we need to decide whether to re-shuffle the smaller datasets mid-epoch when they are over or restart drawing batches in the same order until the largest dataset is over.

I think I'd be in favour of a single batch size for all datasets and option 2 without multiple reshuffling, but I'm very open to suggestions or alternatives.

**Comment by luigibonati** [comment]
I would say to start by keeping the number of batches the same for all datasets (instead of the size of the batches). all other scenarios require to make a choice whose effects need to be tested, so i would avoid doing it now.

by default the dataloaders have a batch size equal to the lenght so we can just monitor the number of items in the dataloader and throw and error explaining it if they are different.

what do you think?

**Comment by andrrizzi** [comment]
Yeah, that might be the easiest way to enable datasets of different lengths.

Since ``DictDataLoader`` takes ``batch_size`` as input (not ``n_batches``) I'll make it mean the batch size of the first dataset and set automatically the others based on the number of batches in the first dataset.

**Comment by andrrizzi** [comment]
Hmm. The only problem is that different batch sizes for the datasets will result in difference variance for the different loss functions so I guess the effect of this also has to be tested, but at least I think it'll be easier to implement.

**Comment by luigibonati** [comment]
> Yeah, that might be the easiest way to enable datasets of different lengths.
> 
> Since `DictDataLoader` takes `batch_size` as input (not `n_batches`) I'll make it mean "the batch size of the first dataset" and set automatically the others based on the number of batches in the first dataset.

sounds good to me



> Hmm. The only problem is that different batch sizes for the datasets will result in difference variance for the different loss functions so I guess the effect of this also has to be tested, but at least I think it'll be easier to implement.

I agree, but that will depend on the specific loss functions used right? so I would add a warning  in the tutorials for the multi-task learning to check for this aspect
